src/vm/peimage.cpp

```diff
@@ -1104,7 +1104,10 @@ void PEImage::Load()
 
     SimpleWriteLockHolder lock(m_pLayoutLock);
 
-    _ASSERTE(m_pLayouts[IMAGE_LOADED] == NULL);
+    if (m_pLayouts[IMAGE_LOADED] != NULL)
+    {
+        return;
+    }
```

Comment:
@jkotas, Thank you for explanation.
So, the `Volatile<LONG> m_RWLock;` in `SimpleRWLock` actually have barrier semantics. I incorrectly considered it as standard `volatile` value, while actually it is special implementation with additional barrier semantics.